### PR TITLE
Backport Algorand v2 -> v1

### PR DESCRIPTION
**Pull request by haammar-ledger**
Draft PR just to review the changes.

### CODE REVIEW

File: core/src/wallet/algorand/database/AlgorandAccountDatabaseHelper.hpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+#ifndef LEDGER_CORE_ALGORANDACCOUNTDATABASEHELPER_H
+#define LEDGER_CORE_ALGORANDACCOUNTDATABASEHELPER_H
+
+#include "../model/AlgorandAccount.hpp"
+
+#include "../operations/AlgorandOperation.hpp"
+
+#include <wallet/common/AbstractAccount.hpp>
```
Comment (haammar-ledger) [inline core/src/wallet/algorand/database/AlgorandAccountDatabaseHelper.hpp:37]: A bit of cleanup opportunity: all these can be removed or moved to the `.cpp`

File: core/src/wallet/algorand/AlgorandWalletFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+#include "AlgorandWalletFactory.hpp"
+#include "AlgorandWallet.hpp"
+#include "AlgorandNetworks.hpp"
+
+#include <api/AlgorandConfigurationDefaults.hpp>
+#include <api/AlgorandBlockchainExplorerEngines.hpp>
+#include <api/AlgorandBlockchainObserverEngines.hpp>
+
+//#include <core/Services.hpp>
+#include <wallet/pool/WalletPool.hpp> // V1 for core/Services.hpp ?
```
Comment (haammar-ledger) [inline core/src/wallet/algorand/AlgorandWalletFactory.cpp:40]: Already included in `.hpp`

File: core/src/wallet/algorand/AlgorandModelMapper.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
+/*
+ * AlgorandModelMapper
```
Comment (haammar-ledger) [inline core/src/wallet/algorand/AlgorandModelMapper.cpp:2]: This file belongs in core/src/wallet/algorand/**model**/

File: core/src/wallet/algorand/AlgorandLikeCurrencies.hpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+namespace ledger {
+    namespace core {
+                namespace algorand {
+            namespace constants {
+                static constexpr uint64_t COIN_ID = 283;
+            }
+        }
+        namespace currencies {
+            extern LIBCORE_EXPORT const api::Currency ALGORAND;
+        }
```
Comment (haammar-ledger) [inline core/src/wallet/algorand/AlgorandLikeCurrencies.hpp:52]: This is now defined in `currencies.hpp`, you can remove it from here, or comment it out if you prefer.

File: core/src/wallet/algorand/AlgorandAccountSynchronizer.hpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+#ifndef LEDGER_CORE_ALGORANDACCOUNTSYNCHRONIZER_H
+#define LEDGER_CORE_ALGORANDACCOUNTSYNCHRONIZER_H
+
+#include "AlgorandAddress.hpp"
+#include "AlgorandBlockchainExplorer.hpp"
+
+//#include <core/Services.hpp>
+
+#include <wallet/pool/WalletPool.hpp> // V1 for core/Services.hpp ?
```
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply core/src/wallet/algorand/AlgorandAccountSynchronizer.hpp:39]: (and everywhere else)

File: core/src/wallet/algorand/AlgorandModelMapper.cpp (hunk 3 above)
Comment (barjonr) [reply core/src/wallet/algorand/AlgorandModelMapper.cpp:2]: It's actually my mistake, in V2, the hpp is in model but not the cpp, I'll also change it then

File: core/src/wallet/algorand/AlgorandAccount.cpp [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+    }
+
+    void Account::hasAsset(const std::string & addr, const std::string & assetId, const std::shared_ptr<api::BoolCallback> & callback)
+    {
+        _explorer->getAccount(addr)
+            .map<bool>(
+                    getContext(),
+                    [&assetId](const model::Account& account) {
+                        const auto id = std::stoull(assetId);
+                        return static_cast<bool>(account.assetsAmounts.find(id)->second.amount);
```
Comment (barjonr) [inline core/src/wallet/algorand/AlgorandAccount.cpp:186]: This is not correct. If the account does not hold the asset, you will still be calling `->second.amount` on the result of `account.assetsAmounts.find(id)`, which will be `std::end(account.assetsAmounts)` causing undefined behaviour.
In addition, you can have a total of `0` for the asset and still want to return `true` (assuming the goal of this function is to know whether or not a user can send an ASA to another user).
Your return statement should be `return account.assetsAmounts.count(id) > 0;` or `return account.assetsAmounts.find(id) != std::end(account.assetsAmounts);`.

Comment (haammar-ledger) [reply core/src/wallet/algorand/AlgorandAccount.cpp]: Good catch!

**Pull request merged**